service/src/event/service.rs

```diff
@@ -129,23 +129,72 @@ impl CeramicEventService {
         )?)
     }
 
-    pub(crate) async fn insert_events(
-        &self,
-        items: &[recon::ReconItem<EventId>],
-        source: DeliverableRequirement,
-    ) -> Result<InsertResult> {
-        let mut to_insert = Vec::new();
-        let mut invalid = Vec::new();
+    async fn validate_signed_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    async fn validate_time_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    pub(crate) async fn validate_events(
+        items: &[ReconItem<EventId>],
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        let mut parsed_events = Vec::new();
+        let mut invalid_events = Vec::new();
         for event in items {
             match Self::parse_discovered_event(event).await {
-                Ok(insertable) => to_insert.push(insertable),
-                Err(err) => invalid.push(InvalidItem::InvalidFormat {
+                Ok(insertable) => parsed_events.push(insertable),
+                Err(err) => invalid_events.push(InvalidItem::InvalidFormat {
                     key: event.key.clone(),
                     reason: err.to_string(),
                 }),
             }
         }
 
+        // Group events by their type
+        let mut valid_events = Vec::new();
+        let mut signed_events = Vec::new();
+        let mut time_events = Vec::new();
+        for event in parsed_events {
+            match event.event() {
+                Event::Time(_) => {
+                    time_events.push(event);
+                }
+                Event::Signed(_) => {
+                    signed_events.push(event);
+                }
+                Event::Unsigned(_) => {
+                    // Unsigned events need no extra validation.
+                    valid_events.push(event);
+                }
+            }
+        }
+
+        let (valid_signed, invalid_signed) = Self::validate_signed_events(signed_events).await?;
```

Comment:
maybe use try_join! on the signed/time event futures?

Reply:
ooh yes good call, updated